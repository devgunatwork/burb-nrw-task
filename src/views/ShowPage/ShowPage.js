import React, { useEffect } from 'react';
import { batch, useDispatch, useSelector } from 'react-redux';
import { useParams } from 'react-router-dom';
import Typography from '@material-ui/core/Typography';
import Error from '../../components/Error';
import Loader from '../../components/Loader';
import Details from '../../components/Details';
import EpisodeList from '../../components/EpisodeList';
import SplitPageLayout from '../../components/SplitPageLayout';
import checkImageUrl from '../../utils/checkImageUrl';
import { FAILED, LOADING } from '../../constants/apiState.constants';
import { SHOW_FETCH_REQUESTED } from '../../state/types/showSaga.type';
import { EPISODE_LIST_FETCH_REQUESTED } from '../../state/types/episodeListSaga.type';

/**
 * @description component is container for show page
 * */
const ShowPage = () => {
  const {
    show: { details, status: showRequestStatus },
    episodeList: { list },
  } = useSelector((state) => ({
    show: state.show,
    episodeList: state.episodeList,
  }));
  const { image, name, summary } = details;
  const url = checkImageUrl(image);
  const { showId } = useParams();
  const dispatch = useDispatch();

  useEffect(() => {
    batch(() => {
      dispatch({ type: SHOW_FETCH_REQUESTED, payload: { showId } });
      dispatch({ type: EPISODE_LIST_FETCH_REQUESTED, payload: { showId } });
    });
  }, [showId, dispatch]);

  if (showRequestStatus === LOADING) return <Loader />;
  if (showRequestStatus === FAILED) return <Error message="Failed to fetch show details" />;
  return (
    <SplitPageLayout imageUrl={url}>
      <Details name={name} summary={summary} id={showId} />
      <Typography variant="h5">EPISODE LIST:</Typography>
      <EpisodeList episodes={list} />
    </SplitPageLayout>
  );
};

export default ShowPage;
