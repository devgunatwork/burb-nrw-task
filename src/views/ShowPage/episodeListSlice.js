import { createSlice } from '@reduxjs/toolkit';
import { FAILED, IDLE, LOADING, SUCCEEDED } from '../../constants/apiState.constants';

export const initialState = {
  list: [],
  status: IDLE,
  error: '',
};

export const episodeListSlice = createSlice({
  name: 'episodeList',
  initialState,
  reducers: {
    requested: (state) => {
      state.status = LOADING;
    },
    succeeded: (state, action) => {
      state.status = SUCCEEDED;
      state.list = action.payload.episodeList;
    },
    failed: (state, action) => {
      state.status = FAILED;
      state.error = action.error;
    },
  },
});

export const { requested, succeeded, failed } = episodeListSlice.actions;

export default episodeListSlice.reducer;
