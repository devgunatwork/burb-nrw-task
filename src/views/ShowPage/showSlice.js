import { createSlice } from '@reduxjs/toolkit';
import { FAILED, IDLE, LOADING, SUCCEEDED } from '../../constants/apiState.constants';

export const initialState = {
  details: {},
  status: IDLE,
  error: '',
};

export const showSlice = createSlice({
  name: 'show',
  initialState,
  reducers: {
    requested: (state) => {
      state.status = LOADING;
    },
    succeeded: (state, action) => {
      state.status = SUCCEEDED;
      state.details = action.payload.show;
    },
    failed: (state, action) => {
      state.status = FAILED;
      state.error = action.error;
    },
  },
});

export const { increment, decrement, incrementByAmount } = showSlice.actions;

export const selectCount = (state) => state.counter.value;

export default showSlice.reducer;
