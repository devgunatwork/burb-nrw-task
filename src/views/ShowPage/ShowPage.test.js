import React from 'react';
import { useSelector } from 'react-redux';
import ShowPage from './ShowPage';
import { initialState } from './showSlice';

jest.mock('react-router-dom', () => ({
  ...jest.requireActual('react-router-dom'),
  useParams: () => ({
    id: '1234',
  }),
}));
jest.mock('react-redux', () => ({
  useDispatch: jest.fn(),
  useSelector: jest.fn(),
}));

describe('views', () => {
  describe('ShowPage', () => {
    describe('UI', () => {
      it('should match the snapshot when loading', () => {
        useSelector.mockImplementation((selector) => selector({ show: initialState }));
        const wrapper = shallow(<ShowPage />);

        expect(wrapper).toMatchSnapshot();
      });
    });
  });
});
