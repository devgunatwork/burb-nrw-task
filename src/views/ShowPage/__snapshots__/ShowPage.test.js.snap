// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`views ShowPage UI should match the snapshot when loading 1`] = `
<SplitPageLayout
  imageUrl="/assets/images/404.png"
>
  <Details
    name=""
    summary=""
  />
  <WithStyles(ForwardRef(Typography))
    variant="h5"
  >
    EPISODE LIST:
  </WithStyles(ForwardRef(Typography))>
  <EpisodeList
    episodes={Array []}
  />
</SplitPageLayout>
`;
